### PR TITLE
WIP FileDB Rewrite

### PR DESCRIPTION
**Pull request by Apollon77**
As basis to migrate communication to Redis protocol I did a rewrite of the InMemDBs for states and objects. I decided to use class approach. Here is the result.

* The class "inMemFileDB" is the parent of the class structure and provides basic json storage and general subscription and publish functionality. So this basically removes duplicated code out of the other classes.
* "objectsInMemDB"/"statesInMemDB" classes inherit from the base class to add state/object related functionality, also provides the local interface of the FileDB used by js-controller
* "objectsInMemServer"/"statesInMemServer" inherit from the them then to add socket.io based communication handling to provide a server for remote usage by adapters

So basic functions re now separated from the business functions for states/objects and from communication layer.

Basically I tested the best I could and also trvis is passing but deeper testing may be needed.

In fact one topic to discuss is the question on "private/protected" methods: The old way of writing provided a way to have private methods and so only the allowed methods and properties/variables were pulicly accessable. This is not possible that easy with JS classes anymore :-(
Means: In this PR also only internally available methods like _getObject (which basically would bypass right checking) are generally accessible by js-controller or an adapter. Is this acceptable?!
An alternative would be to use Symbols (https://medium.com/front-end-weekly/private-methods-in-es6-and-writing-your-own-db-b2e30866521f).

Let's discuss :-)

### CODE REVIEW

File: lib/inMemFileDB.js [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+                fs.mkdirSync(this.backupDir);
+            }
+        }
+
+        this.log = utils.getLogger(this.settings.logger);
+
+        // load values from file
+        if (fs.existsSync(this.datasetName)) {
+            try {
+                this.dataset = JSON.parse(fs.readFileSync(this.datasetName).toString());
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: `this.dataset = JSON.parse(fs.readFileSync(this.datasetName, 'utf8'));`

File: lib/inMemFileDB.js [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+            } catch (e) {
+                this.log.error(this.namespace + ' Cannot parse ' + this.datasetName + ': ' + e);
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Time for template strings?
```js
`${this.namespace} Cannot parse ${this.datasetName} ...`
```

File: lib/inMemFileDB.js [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+        } else {
+            this.dataset = {};
+        }
+
+        // Check if directory exists
+        this.datasetName = this.datasetName.replace(/\\/g, '/');
+        /** @type {string|string[]} */
+        let parts = this.datasetName.split('/');
+        parts.pop();
+        parts = parts.join('/');
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: How about just `path.dirname(this.datasetName)`?

File: lib/inMemFileDB.js [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        client._subscribe = client._subscribe || {};
+        if (!client._subscribe[type]) client._subscribe[type] = [];
+        const s = client._subscribe[type];
+
+        if (pattern instanceof Array) {
+            pattern.forEach(pattern => {
+                for (let i = 0; i < s.length; i++) {
+                    if (s[i].pattern === pattern) return;
+                }
+
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Instead of the loop:
`if (s.find(sub => sub.pattern === pattern)) return;`

File: lib/inMemFileDB.js [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+                s.push({pattern: pattern, regex: new RegExp(tools.pattern2RegEx(pattern)), options: options});
+            });
+        } else {
+            for (let i = 0; i < s.length; i++) {
+                if (s[i].pattern === pattern) {
+                    typeof cb === 'function' && cb();
+                    return;
+                }
+            }
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Instead of the loop:
```js
if (s.find(sub => sub.pattern === pattern)) {
typeof cb === 'function' && cb();
return;
}
```

File: lib/inMemFileDB.js [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+        }
+        const s = client._subscribe[type];
+        if (pattern instanceof Array) {
+            pattern.forEach(pattern => {
+                for (let i = 0; i < s.length; i++) {
+                    if (s[i].pattern === pattern) {
+                        s.splice(i, 1);
+                        return;
+                    }
+                }
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Instead of the loop:
```js
const index = s.findIndex(sub => sub.pattern === pattern);
if (index > -1) {
  s.splice(i, 1);
  return;
}
```

File: lib/inMemFileDB.js [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+            });
+        } else {
+            for (let i = 0; i < s.length; i++) {
+                if (s[i].pattern === pattern) {
+                    s.splice(i, 1);
+                    if (typeof cb === 'function') cb();
+                    return;
+                }
+            }
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: dito, but with the callback

File: lib/inMemFileDB.js (hunk 6 above)
Comment (AlCalzone) [reply lib/inMemFileDB.js]: And I think the loop had to be backwards for that to work.

File: lib/inMemFileDB.js [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+        // delete files only if settings.backupNumber is not 0
+        const files = fs.readdirSync(this.backupDir);
+        files.sort();
+        const limit = Date.now() - this.settings.backup.hours * 3600000;
+
+        for (let f = files.length - 1; f >= 0; f--) {
+            if (!files[f].endsWith(this.settings.fileDB.fileName + '.gz')) {
+                files.splice(f, 1);
+            }
+        }
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Instead of the loop:
`files = files.filter(f => !f.endsWith(this.settings.fileDB.fileName + '.gz'));`

File: lib/inMemFileDB.js [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+                try {
+                    fs.unlinkSync(this.backupDir + file);
+                } catch (e) {
+                    this.log.error(`Cannot delete file "${this.backupDir + file}: ${JSON.stringify(e)}`);
+                }
+            }
+        }
+    }
+
+    getTimeStr(date) {
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Isn't that defined similarly somewhere in formatDate?

File: lib/states/statesInMemFileDB.js [hunk 10]
```diff
@@ -0,0 +1,576 @@
...
+    delMessage(id, messageId, callback) {
+        if (this.messagebox[id]) {
+            let found = false;
+            for (let i = this.messagebox[id].length - 1; i >= 0; i--) {
+                if (this.messagebox[id][i]._id === messageId) {
+                    this.messagebox[id].splice(i, 1);
+                    found = true;
+                    break;
+                }
+            }
```
Comment (AlCalzone) [inline lib/states/statesInMemFileDB.js]: This loop is probably cleaner with `this.messagebox[id].findIndex` (see other comments above)

File: lib/states/statesInMemFileDB.js [hunk 11]
```diff
@@ -0,0 +1,576 @@
...
+//    auth: null, //unused
+//    secure: true/false,
+//    certificates: as required by createServer
+//    port: 9000,
+//    host: localhost
+// };
+//
+
+/** @class */
+class StatesInMemoryFileDB extends InMemoryFileDB {
```
Comment (AlCalzone) [inline lib/states/statesInMemFileDB.js]: Same here... What functionality does this add/provide?

File: lib/states/statesInMemFileDB.js [hunk 12]
```diff
@@ -0,0 +1,576 @@
...
+
+    constructor(settings) {
+        settings.fileDB = {
```
Comment (AlCalzone) [inline lib/states/statesInMemFileDB.js]: Are we sure `settings` is not undefined?

File: lib/objects/objectsInMemFileDB.js [hunk 13]
```diff
@@ -0,0 +1,2301 @@
...
+const fs                = require('fs');
+const InMemoryFileDB    = require('../inMemFileDB');
+const tools             = require('../tools.js');
+const utils             = require('./objectsUtils');
+
+/** @class */
+class ObjectsInMemoryFileDB extends InMemoryFileDB {
+
+    constructor(settings) {
+        settings.fileDB = {
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: Check if settings is defined?

File: lib/objects/objectsInMemFileDB.js [hunk 14]
```diff
@@ -0,0 +1,2301 @@
...
+        }
+        this.fileOptions = {};
+        this.files = {};
+        this.writeTimer = null;
+        this.writeIds = [];
+        this.preserveSettings = [];
+        this.defaultNewAcl = this.settings.defaultNewAcl || null;
+        this.namespace = this.settings.namespace || this.settings.hostname || '';
+
+        this.objectsDir = this.dataDir + 'files/';
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.join

File: lib/objects/objectsInMemFileDB.js [hunk 15]
```diff
@@ -0,0 +1,2301 @@
...
+
+        // init default new acl
+        if (this.dataset['system.config'] && this.dataset['system.config'].common && this.dataset['system.config'].common.defaultNewAcl) {
+            this.defaultNewAcl = JSON.parse(JSON.stringify(this.dataset['system.config'].common.defaultNewAcl));
+        }
+    }
+
+    // -------------- FILE FUNCTIONS -------------------------------------------
+    // memServer specific function
+    mkpathSync(rootpath, dirpath) {
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: I think we're using `fs-extra` somewhere already. So we could replace this method with `ensureDir[Sync]`

File: lib/objects/objectsInMemFileDB.js [hunk 16]
```diff
@@ -0,0 +1,2301 @@
...
+
+        if (this.writeTimer) clearTimeout(this.writeTimer);
+
+        // if store immediately
+        if (force) {
+            this.writeTimer = null;
+            // Store dirs description
+            for (let _id = 0; _id < this.writeIds.length; _id++) {
+                try {
+                    fs.writeFileSync(this.objectsDir + this.writeIds[_id] + '/_data.json', JSON.stringify(this.fileOptions[this.writeIds[_id]]));
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.join

File: lib/objects/objectsInMemFileDB.js [hunk 17]
```diff
@@ -0,0 +1,2301 @@
...
+                    this.log.error(this.namespace + ' Cannot write files: ' + this.objectsDir + this.writeIds[_id] + '/_data.json: ' + e.message);
+                }
+            }
+            this.writeIds = [];
+        } else {
+            this.writeTimer = setTimeout(() => {
+                // Store dirs description
+                for (let id = 0; id < this.writeIds.length; id++) {
+                    try {
+                        fs.writeFileSync(this.objectsDir + this.writeIds[id] + '/_data.json', JSON.stringify(this.fileOptions[this.writeIds[id]]));
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: dito

File: lib/objects/objectsInMemFileDB.js [hunk 18]
```diff
@@ -0,0 +1,2301 @@
...
+
+    insert(id, attName, ignore, options, obj, callback) {
+        return utils.insert(this, id, attName, ignore, options, obj, callback);
+    }
+
+    _writeFile(id, name, data, options, callback) {
+        try {
+            try {
+                if (!fs.existsSync(this.objectsDir))      fs.mkdirSync(this.objectsDir);
+                if (!fs.existsSync(this.objectsDir + id)) fs.mkdirSync(this.objectsDir + id);
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.join

File: lib/objects/objectsInMemFileDB.js [hunk 19]
```diff
@@ -0,0 +1,2301 @@
...
+            } catch (e) {
+                this.log.error(this.namespace + ' Cannot create directories: ' + this.objectsDir + id + ': ' + e.message);
+                this.log.error(this.namespace + ' Check the permissions! Or call "sudo chmod -R 774 *" in ' + tools.appName +' dir');
+                if (typeof callback === 'function') callback(e.message);
+                return;
+            }
+
+            const ext         = name.match(/\.[^.]+$/);
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.extname?

File: lib/objects/objectsInMemFileDB.js [hunk 20]
```diff
@@ -0,0 +1,2301 @@
...
+            if (isBinary) {
+                // Reload by read
+                delete this.files[id][name];
+            } else {
+                this.files[id][name] = data;
+            }
+
+            try {
+                // Create directories if complex structure
+                this.mkpathSync(this.objectsDir + id + '/', name);
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: fs-extra -> ensureDirSync? and also use path.join

File: lib/objects/objectsInMemFileDB.js [hunk 21]
```diff
@@ -0,0 +1,2301 @@
...
+            });
+        }
+
+        const _path = utils.sanitizePath(id, name, callback);
+        id = _path.id;
+        name = _path.name;
+
+        try {
+            if (!this.fileOptions[id]) {
+                if (fs.existsSync(this.objectsDir + id + '/_data.json')) {
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.join

Comment (AlCalzone) [reply lib/objects/objectsInMemFileDB.js]: And maybe we can store that in a variable, since it is used a few times.

File: lib/objects/objectsInMemFileDB.js [hunk 22]
```diff
@@ -0,0 +1,2301 @@
...
+                                owner:       (this.defaultNewAcl && this.defaultNewAcl.owner)            || utils.CONSTS.SYSTEM_ADMIN_USER,
+                                ownerGroup:  (this.defaultNewAcl && this.defaultNewAcl.ownerGroup)       || utils.CONSTS.SYSTEM_ADMIN_GROUP,
+                                permissions: (this.defaultNewAcl && this.defaultNewAcl.file.permissions) || (utils.CONSTS.ACCESS_USER_ALL | utils.CONSTS.ACCESS_GROUP_ALL | utils.CONSTS.ACCESS_EVERY_ALL) // 777
+                            }
+                        };
+                    }
+
+                    this.files[id][name] = fs.readFileSync(this.objectsDir + id + '/' + name);
+                    if (this.fileOptions[id][name].binary === undefined) {
+                        const pos = name.lastIndexOf('.');
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.extname?

File: lib/objects/objectsInMemFileDB.js [hunk 23]
```diff
@@ -0,0 +1,2301 @@
...
+                            permissions: (this.defaultNewAcl && this.defaultNewAcl.file)       || (utils.CONSTS.ACCESS_USER_RW | utils.CONSTS.ACCESS_GROUP_READ | utils.CONSTS.ACCESS_EVERY_READ) // '0644'
+                        };
+                    }
+
+                    this.fileOptions[id][f].acl.permissions = options.mode;
+                    if (fs.existsSync(this.objectsDir + id + '/' + f)) {
+                        const stats = fs.statSync(this.objectsDir + id + '/' + f);
+                        const acl = this.fileOptions[id][f];
+                        const parts = f.split('/');
+                        const fileName = parts.pop();
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.basename?

File: lib/objects/objectsInMemFileDB.js [hunk 24]
```diff
@@ -0,0 +1,2301 @@
...
+                        processed.push({
+                            path:       parts.join('/'),
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: path.dirname?

File: lib/objects/objectsInMemFileDB.js [hunk 25]
```diff
@@ -0,0 +1,2301 @@
...
+        });
+    }
+
+    // -------------- OBJECT FUNCTIONS -------------------------------------------
+    clone(obj) {
+        if (obj === null || obj === undefined || typeof obj !== 'object') {
+            return obj;
+        }
+
+        const temp = obj.constructor(); // changed
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: can we do that without `new`?

File: lib/objects/objectsInMemFileDB.js [hunk 26]
```diff
@@ -0,0 +1,2301 @@
...
+'use strict';
+
+const extend            = require('node.extend');
+const fs                = require('fs');
+const InMemoryFileDB    = require('../inMemFileDB');
+const tools             = require('../tools.js');
+const utils             = require('./objectsUtils');
+
+/** @class */
+class ObjectsInMemoryFileDB extends InMemoryFileDB {
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: Add JSDoc comment describing what this adds.

Also I scrolled quickly thorugh this file as it is huge. So don't expect too much

File: lib/states/statesInMemServer.js [hunk 27]
```diff
@@ -1,1032 +1,267 @@
...
+
+    publishToClients(client, type, id, obj) {
+        if (!client._subscribe || !client._subscribe[type]) return;
+        const s = client._subscribe[type];
+        for (let i = 0; i < s.length; i++) {
+            if (s[i].regex.test(id)) {
+                client.emit('message', s[i].pattern, id, obj);
+                return;
+            }
+        }
```
Comment (AlCalzone) [inline lib/states/statesInMemServer.js]: Can be replaced with `s[i].find(...)`

File: lib/states/statesInMemServer.js [hunk 28]
```diff
@@ -1,1032 +1,267 @@
...
+    initSocket(socket) {
+        if (this.settings.auth) {
+            const user = null;
+            this.socketEvents(socket, user);
+        } else {
+            this.socketEvents(socket);
+        }
+    }
+
+    _initWebServer(settings, server) {
```
Comment (AlCalzone) [inline lib/states/statesInMemServer.js]: Isn't this duplicated between the two MemServer classes? Could be partially extracted into another module.

File: lib/inMemFileDB.js [hunk 29]
```diff
@@ -0,0 +1,346 @@
...
+                }
+            }
+        }
+        if (typeof cb === 'function') cb();
+
+    }
+
+    publishToClients(client, type, id, obj) {
+        throw new Error('no communication handling implemented');
+    }
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Does it make sense to define an "abstract" method here if it can only be used after being overwritten?

File: lib/inMemFileDB.js [hunk 30]
```diff
@@ -0,0 +1,346 @@
...
+        this.change = settings.change;
+
+        this.dataset = {};
+
+        this.namespace = this.settings.namespace || '';
+        this.lastSave = null;
+        this.zlib = null;
+        this.callbackSubscriptionClient = {};
+
+        this.settings.backup = this.settings.backup || {
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply lib/inMemFileDB.js]: I think assumption is that it is set completely from iobroker-data/iobroker.json or not from older installations, so no "missing peroperty extension"

File: lib/inMemFileDB.js [hunk 31]
```diff
@@ -0,0 +1,346 @@
...
+        };
+
+        // path is always relative to appName.js-controller
+        this.dataDir = (this.settings.connection.dataDir || getDefaultDataDir());
+        if (this.dataDir) {
+            if (this.dataDir[0] === '.' && this.dataDir[1] === '.') {
+                this.dataDir = __dirname + '/../' + this.dataDir;
+            } else if (this.dataDir[0] === '.' && this.dataDir[1] === '/') {
+                this.dataDir = __dirname + '/../' + this.dataDir.substring(2);
+            }
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply lib/inMemFileDB.js]: yes

File: lib/inMemFileDB.js (hunk 8 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: great idea

File: lib/inMemFileDB.js [hunk 32]
```diff
@@ -0,0 +1,346 @@
...
+        }
+        this.dataDir = path.normalize(this.dataDir);
+        this.dataDir = this.dataDir.replace(/\\/g, '/');
+        if (this.dataDir[this.dataDir.length - 1] !== '/') this.dataDir += '/';
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply lib/inMemFileDB.js]: Not if a path is coming from settings like this.settings.connection.dataDir ... or?!

File: lib/inMemFileDB.js (hunk 5 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: I solved it differently but comparable

File: lib/inMemFileDB.js (hunk 9 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: but formatDate is adapter class only ...

File: lib/states/statesInMemFileDB.js (hunk 10 above)
Comment (Apollon77) [reply lib/states/statesInMemFileDB.js]: solution is easier ... methods only needed as stubs for compat reason. noone writes into this.messagebox at the moment :-(
I commented out all code (but left it there)

File: lib/objects/objectsInMemFileDB.js (hunk 15 above)
Comment (Apollon77) [reply lib/objects/objectsInMemFileDB.js]: js.controller do not have fs-extra as dep

File: lib/objects/objectsInMemFileDB.js (hunk 26 above)
Comment (Apollon77) [reply lib/objects/objectsInMemFileDB.js]: I did deeper check for the basic things you noticed

File: lib/states/statesInMemServer.js (hunk 28 above)
Comment (Apollon77) [reply lib/states/statesInMemServer.js]: Yes it is the same ... but we do not have "multi inheritence" :-)

yes we could extract into a "socketioTools.js" or such and put init and destroy and maybe also publishToClients there ...  is it really worth it?

File: lib/inMemFileDB.js (hunk 29 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: In fact "throw new error" is best practice at least what I found

File: lib/inMemFileDB.js (hunk 2 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: honestly: too much effort for now :-)

File: lib/objects/objectsInMemFileDB.js (hunk 25 above)
Comment (Apollon77) [reply lib/objects/objectsInMemFileDB.js]: I have no idea .. I only know that it works that way all the time ...

File: lib/states/statesInMemServer.js (hunk 28 above)
Comment (AlCalzone) [reply lib/states/statesInMemServer.js]: I'm just talking about initWebServer. The implementation is pretty specific and does not really need `this`, except for logging and 1 or 2 assignments. Its just something one might forget when updating one class, but not the other.

File: lib/inMemFileDB.js (hunk 29 above)
Comment (AlCalzone) [reply lib/inMemFileDB.js]: I'm not sure if this even belongs into the DB class when it has no implementation and is only needed in the Server classes.

File: lib/inMemFileDB.js [hunk 33]
```diff
@@ -66,26 +69,18 @@ class InMemoryFileDB {
...
-        if (this.dataDir) {
-            if (this.dataDir[0] === '.' && this.dataDir[1] === '.') {
-                this.dataDir = __dirname + '/../' + this.dataDir;
-            } else if (this.dataDir[0] === '.' && this.dataDir[1] === '/') {
-                this.dataDir = __dirname + '/../' + this.dataDir.substring(2);
-            }
-        }
-        this.dataDir = path.normalize(this.dataDir);
+        this.dataDir = path.normalize(path.join(__dirname, '..', this.dataDir) + path.sep);
         this.dataDir = this.dataDir.replace(/\\/g, '/');
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Addinig path.sep and replacing \\ with / should also not be necessary if we use the path methods everywhere.

File: lib/inMemFileDB.js [hunk 34]
```diff
@@ -66,26 +69,18 @@ class InMemoryFileDB {
...
         if (!fs.existsSync(this.dataDir)) {
             fs.mkdirSync(this.dataDir);
         }
 
-        this.datasetName = this.dataDir + this.settings.fileDB.fileName;
+        this.datasetName = path.join(this.dataDir, this.settings.fileDB.fileName);
         this.stateTimer = null;
 
-        this.backupDir = this.settings.backup.path || (this.dataDir + this.settings.fileDB.backupDirName + '/');
+        this.backupDir = this.settings.backup.path || (path.join(this.dataDir, this.settings.fileDB.backupDirName) + path.sep);
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Addinig path.sep should not be necessary if we use the path methods everywhere. We should avoid having to manually add slashes at the end of a path - just another source for errors.

File: lib/states/statesInMemServer.js (hunk 28 above)
Comment (Apollon77) [reply lib/states/statesInMemServer.js]: The balance question here is for me "a new file just because one method" ... ;-) But I see your point

File: lib/inMemFileDB.js [hunk 35]
```diff
@@ -144,9 +139,7 @@ class InMemoryFileDB {
         // Check if directory exists
         this.datasetName = this.datasetName.replace(/\\/g, '/');
         /** @type {string|string[]} */
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: No need for a type annotation now

File: lib/inMemFileDB.js (hunk 29 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: Needs to because I also put the whole subscription (and so also publishing) logic in the main class because it is used for all. But how to publish to connected clients is specific to the communication implementation later

File: lib/inMemFileDB.js (hunk 33 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: The topic is that "this.dataDir" is coming from config and so can be any kind of path ... I would leave it as security net

File: lib/inMemFileDB.js (hunk 34 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: path.join will not add tailing path separators, but we want tohave one here ... ok I can check again whey/where

File: lib/inMemFileDB.js [hunk 36]
```diff
@@ -69,7 +69,9 @@ class InMemoryFileDB {
 
         // path is always relative to appName.js-controller
         this.dataDir = (this.settings.connection.dataDir || getDefaultDataDir());
-        this.dataDir = path.normalize(path.join(__dirname, '..', this.dataDir) + path.sep);
+        if (this.dataDir.startsWith('.')) {
```
Comment (AlCalzone) [inline lib/inMemFileDB.js]: Does this check for relative vs absolute paths? Then the check should be `if (!path.isAbsolute(this.dataDir))`. And the comment above is wrong because it states that the path is always relative.

File: lib/objects/objectsInMemFileDB.js [hunk 37]
```diff
@@ -904,11 +904,11 @@ class ObjectsInMemoryFileDB extends InMemoryFileDB {
                             fs.unlinkSync(location);
                         }
                     }
-                    const path = path.dirname(f);
+                    const filepath = path.dirname(f);
```
Comment (AlCalzone) [inline lib/objects/objectsInMemFileDB.js]: Nitpick, but: `filePath` with the same capitalization

File: lib/inMemFileDB.js (hunk 34 above)
Comment (AlCalzone) [reply lib/inMemFileDB.js]: > path.join will not add tailing path separators, but we want tohave one here

I know but I'm arguing we might/should not need it.

Comment (Apollon77) [reply lib/inMemFileDB.js]: changed in next commit

File: lib/inMemFileDB.js (hunk 33 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: PS: original code https://github.com/ioBroker/ioBroker.js-controller/blob/master/lib/objects/objectsInMemServer.js#L44

So I changed it and removed the comment

File: lib/inMemFileDB.js (hunk 36 above)
Comment (Apollon77) [reply lib/inMemFileDB.js]: I hd the case on travis where it was anabsolute path and so it broke and original code checked for .. or ./ on the beginning :-) so es comment is wrong

Comment (Apollon77) [reply lib/inMemFileDB.js]: see https://github.com/ioBroker/ioBroker.js-controller/blob/master/lib/objects/objectsInMemServer.js#L44 ... original code. I changed it and removed the comment

File: lib/objects/objectsInMemFileDB.js (hunk 37 above)
Comment (Apollon77) [reply lib/objects/objectsInMemFileDB.js]: yw ;-)

**Comment by GermanBluefox** [comment]
LGTM :)

**Comment by Apollon77** [comment]
> LGTM :)
heisst die jetzt aufrufbaren private methoden sind in Ordnung? :-)

**Comment by Apollon77** [comment]
Ok, checked with BF again: We do not change back to private methods for now ... in future we need to find ways to better secure adapter.js

**Comment by Apollon77** [comment]
Then I will simply merge this once testing is green again and use this as basis for the redis protocol changes

File: lib/states/statesInMemServer.js (hunk 28 above)
Comment (AlCalzone) [reply lib/states/statesInMemServer.js]: One method used in two separate locations, yes that warrants a new file in my opinion. I'd rather have 200 modules with a clear intent than having to search where a method is also duplicated to :)

File: lib/inMemFileDB.js (hunk 29 above)
Comment (AlCalzone) [reply lib/inMemFileDB.js]: Ok, makes sense! :+1:

**Comment by AlCalzone** [comment]
> in future we need to find ways to better secure adapter.js

We could use a wrapper around the objects an states DB that only exposes the public methods. Could also be done semi-automatically by copying the methods that are not prefixed with _

File: lib/states/statesInMemServer.js (hunk 28 above)
Comment (Apollon77) [reply lib/states/statesInMemServer.js]: Ok, lets do the following:  we revisit that again when I do the redis protocol changes :-)

**Comment by Apollon77** [comment]
in fact this is already encapsulated but e.g. getObjectsView is missing because adapters use it directly  on adapter.objects.getObjectsView. ;-) -.. created #359
